iceoryx2-ffi/cxx/include/iox2/service_builder_publish_subscribe.hpp

```diff
@@ -149,29 +142,81 @@ inline ServiceBuilderPublishSubscribe<Payload, UserHeader, S>::ServiceBuilderPub
 template <typename PayloadType>
-inline auto ServiceBuilderPublishSubscribe<Payload, UserHeader, S>::get_payload_type_name() ->
-    typename std::enable_if_t<!HasPayloadTypeNameMember<PayloadType>::value, const char*> {
+inline auto ServiceBuilderPublishSubscribe<Payload, UserHeader, S>::get_payload_type_name()
+    -> internal::FromNonSlice<PayloadType> {
+    if (std::is_same_v<PayloadType, uint8_t>) {
+        return "u8";
+    }
+    if (std::is_same_v<PayloadType, uint16_t>) {
+        return "u16";
+    }
+    if (std::is_same_v<PayloadType, uint32_t>) {
+        return "u32";
+    }
+    if (std::is_same_v<PayloadType, uint64_t>) {
+        return "u64";
+    }
+    if (std::is_same_v<PayloadType, int8_t>) {
+        return "i8";
+    }
+    if (std::is_same_v<PayloadType, int16_t>) {
+        return "i16";
+    }
+    if (std::is_same_v<PayloadType, int32_t>) {
+        return "i32";
+    }
+    if (std::is_same_v<PayloadType, int64_t>) {
+        return "i64";
+    }
+    if (std::is_same_v<PayloadType, float>) {
+        return "f32";
+    }
+    if (std::is_same_v<PayloadType, double>) {
+        return "f64";
+    }
+    if (std::is_same_v<PayloadType, bool>) {
+        return "bool";
+    }
     return typeid(typename PayloadInfo<PayloadType>::ValueType).name();
 }
+// NOLINTEND(readability-function-size)
+
+template <typename Payload, typename UserHeader, ServiceType S>
+template <typename PayloadType>
+inline auto ServiceBuilderPublishSubscribe<Payload, UserHeader, S>::get_payload_type_name()
+    -> internal::FromSliceWithCustomizedInnerPayloadTypeName<PayloadType> {
+    return PayloadType::ValueType::PAYLOAD_TYPE_NAME;
+}
+
+template <typename Payload, typename UserHeader, ServiceType S>
+template <typename PayloadType>
+inline auto ServiceBuilderPublishSubscribe<Payload, UserHeader, S>::get_payload_type_name()
+    -> internal::FromSliceWithoutCustomizedInnerPayloadTypeName<PayloadType> {
+    return get_payload_type_name<typename PayloadType::ValueType>();
+}
 
 template <typename Payload, typename UserHeader, ServiceType S>
 template <typename UserHeaderType>
 inline auto ServiceBuilderPublishSubscribe<Payload, UserHeader, S>::get_user_header_type_name() ->
-    typename std::enable_if_t<HasUserHeaderTypeNameMember<UserHeaderType>::value, const char*> {
+    typename std::enable_if_t<internal::HasUserHeaderTypeNameMember<UserHeaderType>::value, const char*> {
     return UserHeaderType::USER_HEADER_TYPE_NAME;
 }
 
 template <typename Payload, typename UserHeader, ServiceType S>
 template <typename UserHeaderType>
 inline auto ServiceBuilderPublishSubscribe<Payload, UserHeader, S>::get_user_header_type_name() ->
-    typename std::enable_if_t<!HasUserHeaderTypeNameMember<UserHeaderType>::value, const char*> {
+    typename std::enable_if_t<!internal::HasUserHeaderTypeNameMember<UserHeaderType>::value, const char*> {
+    if (std::is_void_v<UserHeader>) {
+        return "()"; // no user header provided
+    }
```

Comment:
I wonder whether this could also be set directly on the in the ffi crate. I mean in addition, not instead. Just to have this also work for the C binding when no user header is defined.

Reply:
I added this to `iceoryx2-ffi/ffi/src/api/service_builder.rs`.